store: implement check nullifiers handle

hackaugusto opened the pull request:
This is a WIP.

Pending:

- Implement the client on the RPC server
- Add tests

hackaugusto commented on store/src/server/api.rs:
```diff
@@ -1,27 +1,49 @@
+            while let Some(msg) = messages.next().await {
+                let result = if let Some(nullifier) = msg?.nullifier {
+                    let digest = RpoDigest::new([Felt::new(nullifier.d0), Felt::new(nullifier.d1), Felt::new(nullifier.d2), Felt::new(nullifier.d3)]);
+                    let leaf = tree.read().await.get_value(digest);
```
Note: A single leaf is read at the time, meaning a concurrent write may happen. I think this is the correct behavior here, the alternative would be to copy the tree, or to forbid writes while there is a stream on-going.

bobbinth commented on store/src/server/api.rs:
```diff
@@ -1,27 +1,49 @@
+        tokio::spawn(async move {
+            while let Some(msg) = messages.next().await {
+                let result = if let Some(nullifier) = msg?.nullifier {
+                    let digest = RpoDigest::new([Felt::new(nullifier.d0), Felt::new(nullifier.d1), Felt::new(nullifier.d2), Felt::new(nullifier.d3)]);
```
We should probably check here whether `d0`, `d1` etc. are valid field elements (`Felt::new()` will do a silent reduction even for values out of range). Ideally, we probably should add `TryFrom<[u64; 4]>` to `RpoDigest` - but even a helper function here could work fine.

bobbinth commented on store/src/server/api.rs:
```diff
@@ -1,27 +1,104 @@
+                    Felt::new(nullifier.d2),
+                    Felt::new(nullifier.d3),
+                ]);
+                let leaf = tree.read().await.get_value(digest);
+
+                tx.send(Ok(CheckNullifiersResponse {
+                    result: leaf != TieredSmt::EMPTY_VALUE,
+                }))
+                .await
```
Here, we should also be sending a Merkle path from the tree root to this nullifier.

hackaugusto replied:
implemented

bobbinth reviewed (changes requested):
Thank you! Looks good! I left a few small comments inline.

I do have a general question about the approach: would it preserve the "atomicity" of requests? Meaning, if I make a request to check 100 nullifiers, I'd like the responses to be against the same state of the nullifier db. What I'd like to avoid is getting a response for the first 50 nullifiers against some state of the tree, and then a response for another 50 nullifiers against another state of the tree.

With the current approach (assuming I understood it correctly), it seems like we lock the tree for each individual nullifier check, and if the tree gets updated in between these checks, we might the situation I described above.

So, I guess there are two questions:

1. Do we actually need to have atomic requests? (I think we do, but maybe I'm overthinking it).
2. If we do, what's the best way to achieve it? (again, assuming that the current approach doesn't provide this).

Another thing: the user should be able to verify that the node sent correct data. This means we need to change the responses to include:
- Merkle path from the nullifier to the root of the nullifier tree.
- Value of the nullifier - this can be sent as a u32 value as that's what we'll use to track block heights.

hackaugusto commented:
> would it preserve the "atomicity" of requests? 

No, that is what I mentioned on this comment: https://github.com/0xPolygonMiden/miden-node/pull/14#discussion_r1275537823

IMO locking is not a suitable approach here. It would mean it is possible for an attacker to start a request and hold to it, the serve would hang, waiting for the request to complete while holding the lock, starving other requests.

I think this is not an issue, unless you want the merkle proof in the response to be consistent with the user's view. In that case, then this whole approach won't work, because the server only maintains the latest view of the tree, and a user may have any previous version of the tree, and we need to think of an way of querying for nullifier including for a specific version of a tree denominated by its root, which would need multiple versions of the tree in memory (probably can be done with the MerkleStore)

bobbinth commented:
> IMO locking is not a suitable approach here. It would mean it is possible for an attacker to start a request and hold to it, the serve would hang, waiting for the request to complete while holding the lock, starving other requests.

Agreed - but could we not get around it by redefining the endpoint as follows:
```proto
message CheckNullifiersRequest {
    repeated digest.Digest nullifier = 1;
}

message ResponseBytes {
    bytes data = 1;
}

service Api {
    rpc CheckNullifiers(CheckNullifiersRequest) returns (stream ResponseBytes) {}
}
```
Basically, the request is no longer a stream, but the response is still a stream.

> In that case, then this whole approach won't work, because the server only maintains the latest view of the tree, and a user may have any previous version of the tree, and we need to think of an way of querying for nullifier including for a specific version of a tree denominated by its root, which would need multiple versions of the tree in memory (probably can be done with the MerkleStore)

I don't think we should modify the approach too much. The guarantees I'd like to provide are:
- The server returns the data only about the latest state of the nullifier tree - so, no need to keep historical data.
- The "latest state" does not change in the middle of the request - i.e., Merkle paths returned resolve to the same root.

hackaugusto commented:
> Basically, the request is no longer a stream, but the response is still a stream.

that works. why would you like the response to be a stream though?

bobbinth commented:
> why would you like the response to be a stream though?

I thought when we return a lot of data streaming is preferred. But I'm not sure about the tradeoffs (i.e., whether the complexity is worth it).

hackaugusto commented:
> I thought when we return a lot of data streaming is preferred. But I'm not sure about the tradeoffs (i.e., whether the complexity is worth it).

- A digest uses 8 bytes: They are 4 integers encoded as fixed64 (data is expected to be random and won't benefit from compression), that means a single digest uses `8` bytes. 
- Let's assume 4seconds block time (1/3 than ethereum's block time)
- Let's assume 10x Ethereum transactions (10Mi day)
- Let's assume 3 notes consume per transaction (30Mi day)
- That gives 32_850_000_000 nullifiers per year, and a tree of depth 35
- The above means a Tiered SMT of depth 48
- Since we assumed 3 notes per transaction, that means 3 nullifiers checks, each path with 384 bytes for a total of 1152 bytes. Barely a over a KiB, lets say 1.5KiB per response to account for metadata.

For the use case of checking nullifiers prior to running a transaction it will be fine. Is there something else?

bobbinth commented:
Overall, I agree - probably the volume of data for this endpoint is not that big and we may not need to do steaming here. But a couple of notes:

- I'm assuming we'd make requests for more than 3 nullifiers on average in part to hide which nullifiers we are interested in. So, it could be 100 or more nullifiers per request - and we may care only about 10 out of them.
- Not sure about the size of the path yet - it may actually end up being smaller than 384 bytes as we can "compress" paths by not sending roots of empty subtrees (might be good to create a "nice-to-have" issue for this).

Overall, I think this endpoint would probably not return more than 50KB per request (and maybe that's something we could enforce in the future).

hackaugusto commented on rpc/src/server/api.rs:
```diff
@@ -1,36 +1,115 @@
+                    Ok(rpc::CheckNullifiersRequest {
+                        nullifier: Some(nullifier),
+                    }) => Some(nullifier),
+                };
```
done

hackaugusto commented on store/src/server/api.rs:
```diff
@@ -1,27 +1,104 @@
+                    return Err(Status::invalid_argument(
+                        "Digest field d3 is not in the modulos range",
+                    ));
+                }
```
I would have it in every layer, in terms of CPU time, the cost of validation shouldn't be that high. And the development cost of having a bug going unnoticed because we didn't validate the data in some layer is probably much higher.

I added it to the RPC layer too, let me know if you want to change in the DB layer.

hackaugusto commented:
note on the changes:

- RPC changed from returning `Bytes` to return the protobuf message
- Now the RPC endpoints are returning the tiered proof
- added validation for the nullifiers, making sure the serialized values are inside the modulus range
- removed the streaming from both the client and server

hackaugusto commented on proto/build.rs:
```diff
@@ -21,6 +21,7 @@ fn main() -> miette::Result<()> {
     // Generate the stub of the user facing server from its proto file
     tonic_build::configure()
         .file_descriptor_set_path(&file_descriptor_path)
+        .type_attribute(".", "#[derive(Eq, PartialOrd, Ord, Hash)]")
```
`.` means `all generated classes`. So this is adding the derive to all the message types.

hackaugusto commented on proto/proto/tsmt.proto:
```diff
@@ -0,0 +1,38 @@
+
+message Leaf {
+    digest.Digest key = 1;
+    optional digest.Digest value = 2;
```
This is `optional` because that is how empty leaves are encoded (The leaf itself is set to `[0,0,0,0]`, here that would be set on `key`).

bobbinth replied:
I didn't quite follow this part. If the value is `[0, 0, 0, 0]`, do we basically just us a flag rather than storing all 4 zeros? Or is this something else?

hackaugusto replied:
changed, now using the length of the vector as a flag instead.

bobbinth commented on proto/Cargo.toml:
```diff
@@ -11,6 +11,7 @@ rust-version = "1.67"
 
 [dependencies]
 hex = { version = "0.4" }
+miden-crypto = { git = "https://github.com/0xPolygonMiden/crypto", branch = "hacka-tsmt-proof-basic-traits" }
```
Should we re-point this back to `next`?

bobbinth commented on proto/proto/store.proto:
```diff
@@ -1,20 +1,22 @@
 }
 
 message CheckNullifiersResponse {
-    bool result = 1;
+    // Each requested nullifier has its corresponding nullifier proof at the
+    // same position.
+    repeated tsmt.Proof proofs = 1;
 }
```
Should we use a more specialized message here - e.g., something like `NullifierProof`? One reason is that for nullifiers we don't need to user a full word for `value` - just a `u32` should be enough.

hackaugusto replied:
changed

bobbinth commented on proto/proto/tsmt.proto:
```diff
@@ -0,0 +1,38 @@
+//
+// Notes about TSMT:
+// - The tree is a key value store, both are digest values.
+// - The key's most significant byte is used to deterministically assign the
+// pair to one of the tree's entries.
```
nit: this should say "most significant element" or "8 most significant bytes".

Also, minor nit on formatting: I'd indent second lines of each bullet point to make them easier to read. For example:
```Rust
// - The tree is a key value store, both are digest values.
// - The key's most significant byte is used to deterministically assign the
//   pair to one of the tree's entries.
```
Instead of:
```Rust
// - The tree is a key value store, both are digest values.
// - The key's most significant byte is used to deterministically assign the
// pair to one of the tree's entries.
```

hackaugusto replied:
changed

bobbinth commented on rpc/Cargo.toml:
```diff
@@ -14,12 +14,15 @@ anyhow = { version = "1.0" }
 directories = { version = "5.0" }
 figment = { version = "0.10", features = ["toml", "env"] }
+hex = { version = "0.4" }
+miden-crypto = { git = "https://github.com/0xPolygonMiden/crypto", branch = "hacka-tsmt-proof-basic-traits" }
```
Should we revert this back to `next`?

hackaugusto replied:
changed

bobbinth commented on store/Cargo.toml:
```diff
@@ -21,14 +21,15 @@ clap = { version = "4.3" , features = ["derive"] }
 directories = { version = "5.0" }
 figment = { version = "0.10", features = ["toml", "env"] }
-miden-crypto = { version = "0.6.0" }
+miden-crypto = { git = "https://github.com/0xPolygonMiden/crypto", branch = "hacka-tsmt-proof-basic-traits" }
```
Should we revert this back to `next`?

hackaugusto replied:
changed

bobbinth commented on store/Cargo.toml:
```diff
@@ -21,14 +21,15 @@ clap = { version = "4.3" , features = ["derive"] }
 rusqlite_migration = { version = "1.0" }
 serde = { version = "1.0" , features = ["derive"] }
 tokio = { version = "1.29", features = ["rt-multi-thread", "net", "macros"] }
+tokio-stream = { version = "0.1" }
```
Do we still need this?

hackaugusto replied:
removed

bobbinth commented on store/src/server/api.rs:
```diff
@@ -1,27 +1,66 @@
+        request: tonic::Request<store::CheckNullifiersRequest>,
+    ) -> Result<Response<CheckNullifiersResponse>, Status> {
+        // Validate the nullifiers and convert them to RpoDigest values. Stop on first error.
+        let nullifiers = request
+            .into_inner()
+            .nullifiers
+            .into_iter()
+            .map(|v| {
+                v.try_into()
+                    .or(Err(Status::invalid_argument("Digest field is not in the modulos range")))
+            })
+            .collect::<Result<Vec<RpoDigest>, Status>>()?;
```
Probably not for this PR, but rather thoughts on general approach: since the request is coming from an internal component - do we still need do validation? In this case, the overhead is probably minimal and so it is probably better to do this extra checks anyway. But in case validation logic is more complex, we may want to skip it for internal communication.

bobbinth commented on rpc/Cargo.toml:
```diff
@@ -14,12 +14,15 @@ anyhow = { version = "1.0" }
 prost = { version = "0.11" }
 serde = { version = "1.0" , features = ["derive"] }
 tokio = { version = "1.29", features = ["rt-multi-thread", "net", "macros"] }
+tokio-stream = { version = "0.1" }
```
Do we still need this?

hackaugusto replied:
removed

bobbinth reviewed (changes requested):
Looks good! Thank you! I left a some comments inline - most are pretty minor.

hackaugusto commented on proto/proto/tsmt.proto:
```diff
@@ -0,0 +1,41 @@
+
+message NullifierLeaf {
+    digest.Digest key = 1;
+    uint64 value = 2;
```
the comment suggest `u32`, using `u64` because it doesn't requires casting and the zig-zag encoding should result in the same amount of bytes.

bobbinth commented on proto/proto/rpc.proto:
```diff
@@ -1,19 +1,20 @@
+}
+
+message CheckNullifiersResponse {
+    // Each requested nullifier has its corresponding nullifier proof at the
+    // same position.
+    repeated tsmt.NullifierProof proofs = 1;
```
A thought for the future: I wonder if we should also include the block header for block against which we did the check (i.e., the latest block in the chain known to the node).

bobbinth commented on proto/src/conversion.rs:
```diff
@@ -0,0 +1,119 @@
+            .leaves
+            .into_iter()
+            .map(|leaf| {
+                let key = leaf.key.ok_or(error::ParseError::MissingLeafKey)?.try_into()?;
+                let value = [Felt::ZERO, Felt::ZERO, Felt::ZERO, Felt::new(leaf.value)];
+                let result = (key, value);
```
We should probably document this somewhere (not sure where though) - i.e., the fact that the block number is stored in the last element of the leaf.

bobbinth reviewed (approved):
All looks good! Thank you! I left a couple of comments with thoughts for the future.
